### PR TITLE
opt: Add VulkanMemoryModelDeviceScope to trim

### PR DESCRIPTION
**Pull request by sudonatalie**
Add the VulkanMemoryModelDeviceScope capability to the capability trimming pass. According the the spec, "If the Vulkan memory model is declared and any instruction uses Device scope, the VulkanMemoryModelDeviceScope capability must be declared." Since this case, based on the type of an operand, is not covered by the JSON grammar, it is added explicitly.

### CODE REVIEW

File: source/opt/ir_context.h [hunk 1]
```diff
@@ -229,6 +229,8 @@ class IRContext {
   inline void AddExtInstImport(std::unique_ptr<Instruction>&& e);
   // Set the memory model for this module.
   inline void SetMemoryModel(std::unique_ptr<Instruction>&& m);
+  // Get the memory model for this module.
+  inline Instruction* GetMemoryModel();
```
Comment (Keenuts) [inline source/opt/ir_context.h:233]: nit: shall the `const` equivalent of the function be added instead? (or both).

File: test/opt/trim_capabilities_pass_test.cpp [hunk 2]
```diff
@@ -2584,6 +2584,74 @@ TEST_F(TrimCapabilitiesPassTest, UInt16_RemainsWhenUsed) {
...
+       VulkanMemoryModelDeviceScope_RemovedWhenUsedWithGLSL450) {
+  const std::string kTest = R"(
+               OpCapability VulkanMemoryModelDeviceScope
+; CHECK-NOT:   OpCapability VulkanMemoryModelDeviceScope
+               OpCapability Shader
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint GLCompute %main "main"
+               OpExecutionMode %main LocalSize 1 2 4
+       %void = OpTypeVoid
+       %uint = OpTypeInt 32 0
```
Comment (Keenuts) [inline test/opt/trim_capabilities_pass_test.cpp:2617]:
```suggestion
       %uint = OpTypeInt 32 0
      %ulong = OpTypeInt 64 0
```

File: test/opt/trim_capabilities_pass_test.cpp [hunk 3]
```diff
@@ -2584,6 +2584,74 @@ TEST_F(TrimCapabilitiesPassTest, UInt16_RemainsWhenUsed) {
...
+      SinglePassRunAndMatch<TrimCapabilitiesPass>(kTest, /* skip_nop= */ false);
+  EXPECT_EQ(std::get<1>(result), Pass::Status::SuccessWithChange);
+}
+
+TEST_F(TrimCapabilitiesPassTest,
+       VulkanMemoryModelDeviceScope_RemovedWhenUsedWithGLSL450) {
+  const std::string kTest = R"(
+               OpCapability VulkanMemoryModelDeviceScope
+; CHECK-NOT:   OpCapability VulkanMemoryModelDeviceScope
+               OpCapability Shader
```
Comment (Keenuts) [inline test/opt/trim_capabilities_pass_test.cpp:2612]:
```suggestion
               OpCapability Shader
               OpCapability ShaderClockKHR
               OpCapability Int64
               OpExtension "SPV_KHR_shader_clock"
```

**Review by Keenuts — APPROVED** [review]
Just some nits, otherwise LGTM

File: source/opt/ir_context.h (hunk 1 above)
Comment (sudonatalie) [reply source/opt/ir_context.h]: Good point, done

**Review by Keenuts — APPROVED** [review]
Thanks!

**Pull request merged**